src/metabase/email.clj

```diff
@@ -95,12 +98,54 @@
                   (assert (#{:tls :ssl :none :starttls} (keyword new-value))))
                 (setting/set-value-of-type! :keyword :email-smtp-security new-value)))
 
+(defsetting email-rate-limit-per-second
+  (deferred-tru "The maximum number of emails that can be sent per second. The quotas are based on the total number of recipients")
+  :export?    true
+  :type       :integer
+  :visibility :settings-manager
+  :audit      :getter)
+
+(defn- make-email-throttler
+  [rate-limit]
+  (throttle/make-throttler
+   :email
+   :attempt-ttl-ms     1000
+   :initial-delay-ms   1000
+   :attempts-threshold rate-limit))
+
+(defonce ^:private email-throttler (when-let [rate-limit (email-rate-limit-per-second)]
+                                     (make-email-throttler rate-limit)))
+
+(defn check-email-throttle
+  "Check if the email throttler is enabled and if so, throttle the email sending based on the total number of recipients.
+
+  If the number of recipients exceeds the rate limit per second, we max-out the throttle limit but silience the exception.
+  This is to avoid creating regression bugs due to the throttling mechanism."
+  [email]
+  (when email-throttler
+    (let [recipients (set (mapcat #(% email) [:to :bcc]))
+          thorttle-check (fn []
+                           (dotimes [_ (count recipients)]
+                             (throttle/check email-throttler true)))]
+      (if (> (count recipients) (.attempts-threshold ^Throttler email-throttler))
```

Comment:
If I understand the code correctly, we can send an unbounded number of emails as long as they have more recipients than the throttle limit in each message.

Surely we only want to let it through if its the *first* attempt in the current bucket? That minimizes the amount we exceed the limit, while eventually sending everything.

Reply:
the motivation here is that we don't straight out block any emails that have recipients exceed this limit (there is a good chance that the configuration limit is lower than it should be).

I'm not whether we should do this only the first time tho, since there is a good chance that when we let this through, it can still be failed to send due to SES hitting the limit at the time, in this case I think we should let retry mechanism handle it and try sending until it gets through?

Reply:
Trying out some pictures to see if I can explain myself better this way. Here are your current tests in a notation that I hope is clear: (1 emoji = 1 recipient, bracketed per event)

1. (✅)(✅)(✅)(❌)
1. (✅)(❌❌❌)
2. (✅✅✅✅)(❌) - I upped the limit from 2 to 3 for this test for visual consistency

A I understand the current code, we also have the following cases where over-sized events always get through:

1. (✅)(✅)(✅)(✅✅✅✅)(❌)(✅✅✅✅)
1. (✅)(✅✅✅✅)(❌)(✅✅✅✅)

This is very counter intuitive I think, and customers could trip over the fact that setting their threshold lower may actually let more emails through.

The "always let through the first event through, whether or not it breaches" logic I suggest would look like this rather:

1. (✅)(✅)(✅)(❌❌❌❌)(❌)(❌❌❌❌)
1. (✅)(❌❌❌❌)(❌)(❌❌❌❌)

Another slightly more permissive approach for better through put would be to relax this to "allow any event if we haven't breached yet". This would look like the following:

1. (✅)(✅)(✅)(❌❌❌❌)(❌)(❌❌❌❌)
1. (✅)(✅✅✅✅)(❌)(❌❌❌❌)
2. (✅✅✅✅)(❌)

I slightly prefer the former, but think the latter is also fine if you're worried about throughput.

Reply:
We chatted and decided to go with the last suggestion, as it doesn't have that weird edge case where we throttle too much.

Reply:
[09649d9](https://github.com/metabase/metabase/pull/49477/commits/09649d9d4d12c62bc3d5ba7671a0a96f4ff730dc)

the code is uglier than I expected but I think this captures it

Reply:
I think it is overcomplicated now - I don't think you even need to check the number of recipients if you adjust the if check as we discussed.